common/primitives/src/msa.rs

```diff
@@ -295,7 +295,19 @@ pub trait SchemaGrantValidator<BlockNumber> {
 	) -> DispatchResult;
 }
 
-/// RPC Response for getting getting MSA keys
+/// A trait that allows checking whether adding a key would be free
+pub trait MsaKeyProvider {
+	/// the type to use for looking up keys in storage.
+	type AccountId;
+	/// Returns whether adding `new_key` to `msa_id` would be free
+	fn key_eligible_for_free_addition(
```

Comment:
I considered having this accepting `AddKeyData`, but `AddKeyData` takes a `Config` generic, and you wind up with dueling Configs (between frequency-tx-payment and msa).